File: src/Avalonia.Samples/CompleteApps/Avalonia.MusicStore/Services/AlbumService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text.Json;
using System.Threading.Tasks;
using Avalonia.MusicStore.Models;

namespace Avalonia.MusicStore.Services
{
    /// <summary>
    /// Provides album search, cover-art loading, and local cache persistence
    /// by calling the Apple iTunes Web API.
    /// </summary>
    public class AlbumService
    {
        private static readonly HttpClient s_httpClient = new();
        private static readonly JsonSerializerOptions s_jsonOptions = new() { PropertyNameCaseInsensitive = true };

        // Private DTOs for deserializing the iTunes Search API response
        private record ItunesAlbum(string ArtistName, string CollectionName, string ArtworkUrl100);
        private record ItunesSearchResult(ItunesAlbum[] Results);

        private static string CachePath(Album album) =>
            $"./Cache/{SanitizeFileName(album.Artist)} - {SanitizeFileName(album.Title)}";

        /// <summary>
        /// Searches the iTunes API for albums matching the given search term.
        /// </summary>
        public async Task<IEnumerable<Album>> SearchAsync(string? searchTerm)
        {
            if (string.IsNullOrWhiteSpace(searchTerm))
            {
                return Enumerable.Empty<Album>();
            }

            var url = $"https://itunes.apple.com/search?term={Uri.EscapeDataString(searchTerm)}&entity=album&limit=25";
            var json = await s_httpClient.GetStringAsync(url).ConfigureAwait(false);
            var result = JsonSerializer.Deserialize<ItunesSearchResult>(json, s_jsonOptions);

            return result?.Results.Select(x =>
                new Album(x.ArtistName, x.CollectionName,
                    x.ArtworkUrl100.Replace("100x100bb", "600x600bb")))
                ?? Enumerable.Empty<Album>();
        }

        /// <summary>
        /// Loads the cover bitmap for an album from cache or the iTunes API.
        /// </summary>
        public async Task<Stream> LoadCoverBitmapAsync(Album album)
        {
            var cachePath = CachePath(album);
            if (File.Exists(cachePath + ".bmp"))
            {
                return File.Open(cachePath + ".bmp", FileMode.Open, FileAccess.Read, FileShare.Read);
            }
            else
            {
                var data = await s_httpClient.GetByteArrayAsync(album.CoverUrl);
                return new MemoryStream(data);
            }
        }

        /// <summary>
        /// Saves the album metadata to the local cache as a JSON file.
        /// </summary>
        public async Task SaveAsync(Album album)
        {
            if (!Directory.Exists("./Cache"))
            {
                Directory.CreateDirectory("./Cache");
            }

            using var fs = File.OpenWrite(CachePath(album) + ".json");
            await SaveToStreamAsync(album, fs);
        }

        /// <summary>
        /// Opens a write stream for saving the album cover bitmap to cache.
        /// </summary>
        public Stream SaveCoverBitmapStream(Album album)
        {
            return File.OpenWrite(CachePath(album) + ".bmp");
        }

        /// <summary>
        /// Loads all cached albums from the /Cache directory.
        /// </summary>
        public async Task<IEnumerable<Album>> LoadCachedAsync()
        {
            if (!Directory.Exists("./Cache"))
            {
                Directory.CreateDirectory("./Cache");
            }

            var results = new List<Album>();

            foreach (var file in Directory.EnumerateFiles("./Cache"))
            {
                if ((new DirectoryInfo(file).Extension) != ".json")
                    continue;

                await using var fs = File.Open(file, FileMode.Open, FileAccess.Read, FileShare.Read);
                results.Add(await LoadFromStreamAsync(fs).ConfigureAwait(false));
            }

            return results;
        }

        private static async Task<Album> LoadFromStreamAsync(Stream stream)
        {
            return (await JsonSerializer.DeserializeAsync<Album>(stream).ConfigureAwait(false))!;
        }

        private static async Task SaveToStreamAsync(Album data, Stream stream)
        {
            await JsonSerializer.SerializeAsync(stream, data).ConfigureAwait(false);
        }

        /// <summary>
        /// Replaces characters that are invalid in file names with underscores.
        /// Example: AC/DC -> AC_DC
        /// </summary>
        private static string SanitizeFileName(string input)
        {
            foreach (var c in Path.GetInvalidFileNameChars())
            {
                input = input.Replace(c, '_');
            }

            return input;
        }
    }
}

File: src/Avalonia.Samples/FSharp/CompleteApps/MusicStore/Services/AlbumService.fs
namespace MusicStore.Services

open System
open System.IO
open System.Net
open System.Net.Http
open System.Text.Json
open System.Threading.Tasks
open System.Collections.Concurrent
open MusicStore.Models

/// Private DTOs for deserializing the iTunes Search API response
[<CLIMutable>]
type private ItunesAlbum = {
    ArtistName: string
    CollectionName: string
    ArtworkUrl100: string
}

[<CLIMutable>]
type private ItunesSearchResult = {
    Results: ItunesAlbum[]
}

/// Provides album search, cover-art loading, and local cache persistence
/// by calling the Apple iTunes Web API.
type AlbumService() =
    static let s_jsonOptions = JsonSerializerOptions(PropertyNameCaseInsensitive = true)

    // Shared HTTP client for downloading cover images
    static let s_httpClient =
        let handler = new HttpClientHandler()
        handler.AutomaticDecompression <- DecompressionMethods.GZip ||| DecompressionMethods.Deflate
        new HttpClient(handler, disposeHandler = true)

    static do
        s_httpClient.Timeout <- TimeSpan.FromSeconds(15.0)

        try
            s_httpClient.DefaultRequestHeaders.UserAgent.ParseAdd("MusicStore/1.0 (+https://github.com/AvaloniaUI/Avalonia.Samples)")
            s_httpClient.DefaultRequestHeaders.Accept.ParseAdd("image/webp,image/apng,image/*;q=0.8,*/*;q=0.5")
        with exn ->
#if DEBUG
            System.Diagnostics.Debug.WriteLine($"Failed to set HttpClient headers: {exn.ToString()}")
#endif
            ()

    // The `CacheDir` is constructed from the current directory of the spawning process.
    static member private CacheDir = Path.Join(Environment.CurrentDirectory, "Cache")

    /// Replaces characters that are invalid in file names with underscores.
    static member private SanitizeFileName(input: string) =
        Path.GetInvalidFileNameChars()
        |> Array.fold (fun (acc: string) (c: char) -> acc.Replace(c, '_')) input

    member private _.CachePath(album: Album) =
        Path.Join(
            AlbumService.CacheDir,
            $"{AlbumService.SanitizeFileName(album.Artist)} - {AlbumService.SanitizeFileName(album.Title)}"
        )

    /// Searches iTunes for albums matching the given term. Returns an empty sequence for null/whitespace.
    member _.SearchAsync(searchTerm: string option) =
        task {
            let term =
                match searchTerm with
                | Some s -> s
                | None -> ""

            if String.IsNullOrWhiteSpace(term) then
                return Seq.empty<Album>
            else
                let url = $"https://itunes.apple.com/search?term={Uri.EscapeDataString(term)}&entity=album&limit=25"
                let! json = s_httpClient.GetStringAsync(url)
                let result = JsonSerializer.Deserialize<ItunesSearchResult>(json, s_jsonOptions)

                let results =
                    result.Results
                    |> Seq.map (fun x ->
                        let cover = x.ArtworkUrl100.Replace("100x100bb", "600x600bb")
                        new Album(x.ArtistName, x.CollectionName, cover))

                return results
        }

    /// Loads the cover bitmap for an album from cache or the iTunes API.
    member this.LoadCoverBitmapAsync(album: Album) =
        // Download with a simple retry to handle transient CDN hiccups.
        // Must be defined outside the outer task{} to avoid FS3511 (let rec in resumable code).
        let rec download attempt =
            task {
                try
                    let! data = s_httpClient.GetByteArrayAsync(album.CoverUrl)
                    return new MemoryStream(data) :> Stream
                with :? HttpRequestException when attempt < 2 ->
                    do! Task.Delay(200)
                    return! download (attempt + 1)
            }

        task {
            let bmpPath = this.CachePath(album) + ".bmp"

            if File.Exists(bmpPath) then
                return File.OpenRead(bmpPath) :> Stream
            else
                return! download 0
        }

    /// Saves the album metadata to the local cache as a JSON file.
    member this.SaveAsync(album: Album) =
        task {
            if not (Directory.Exists(AlbumService.CacheDir)) then
                Directory.CreateDirectory(AlbumService.CacheDir) |> ignore

            use fs = File.OpenWrite(this.CachePath(album) + ".json")
            do! JsonSerializer.SerializeAsync(fs, album).ConfigureAwait(false)
        }

    /// Opens a write stream for saving the album cover bitmap to cache.
    member this.SaveCoverBitmapStream(album: Album) =
        File.OpenWrite(this.CachePath(album) + ".bmp")

    /// Loads all cached albums from the Cache directory.
    member _.LoadCachedAsync() =
        task {
            if not (Directory.Exists(AlbumService.CacheDir)) then
                Directory.CreateDirectory(AlbumService.CacheDir) |> ignore

            let files =
                Directory.EnumerateFiles(AlbumService.CacheDir, "*.json", SearchOption.TopDirectoryOnly)

            let bag = ConcurrentBag<Album>()
            let opts = ParallelOptions(MaxDegreeOfParallelism = Environment.ProcessorCount)

            do! Parallel.ForEachAsync(
                    files,
                    opts,
                    fun file _ct ->
                        task {
                            use fs = File.OpenRead(file)
                            let! album = JsonSerializer.DeserializeAsync<Album>(fs).ConfigureAwait(false)
                            bag.Add(album)
                        }
                        |> ValueTask
                )

            return seq (bag.ToArray())
        }
